mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala

```diff
@@ -224,142 +261,133 @@ class KMeans private (
-
-    // Only one run is allowed when initialModel is given
-    val numRuns = if (initialModel.nonEmpty) {
-      if (runs > 1) logWarning("Ignoring runs; one run is allowed when initialModel is given.")
-      1
-    } else {
-      runs
-    }
-
-    val centers = initialModel match {
-      case Some(kMeansCenters) => {
-        Array(kMeansCenters.clusterCenters.map(s => new VectorWithNorm(s)))
-      }
-      case None => {
-        if (initializationMode == KMeans.RANDOM) {
-          initRandom(data)
-        } else {
-          initKMeansParallel(data)
-        }
-      }
-    }
-    val initTimeInSeconds = (System.nanoTime() - initStartTime) / 1e9
-    logInfo(s"Initialization with $initializationMode took " + "%.3f".format(initTimeInSeconds) +
-      " seconds.")
-
-    val active = Array.fill(numRuns)(true)
-    val costs = Array.fill(numRuns)(0.0)
-
-    var activeRuns = new ArrayBuffer[Int] ++ (0 until numRuns)
+    var done = false
+    var costs = 0.0
     var iteration = 0
-
     val iterationStartTime = System.nanoTime()
 
-    // Execute iterations of Lloyd's algorithm until all runs have converged
-    while (iteration < maxIterations && !activeRuns.isEmpty) {
+    // Execute Lloyd's algorithm until converged or reached the max number of iterations
+    while (iteration < maxIterations && !done ) {
       type WeightedPoint = (Vector, Long)
       def mergeContribs(x: WeightedPoint, y: WeightedPoint): WeightedPoint = {
         axpy(1.0, x._1, y._1)
         (y._1, x._2 + y._2)
       }
 
-      val activeCenters = activeRuns.map(r => centers(r)).toArray
-      val costAccums = activeRuns.map(_ => sc.accumulator(0.0))
+      val costAccums = sc.accumulator(0.0)
+      val bcCenters = sc.broadcast(centers)
 
-      val bcActiveCenters = sc.broadcast(activeCenters)
 
       // Find the sum and count of points mapping to each center
-      val totalContribs = data.mapPartitions { points =>
-        val thisActiveCenters = bcActiveCenters.value
-        val runs = thisActiveCenters.length
-        val k = thisActiveCenters(0).length
-        val dims = thisActiveCenters(0)(0).vector.size
+      val totalContribs = data.flatMap { case (pointMatrix, pointsNormMatrix) =>
+        val thisCenters = bcCenters.value
+        val k = thisCenters.length
+        val dims = thisCenters(0).vector.size
+
+        val sums = Array.fill(k)(Vectors.zeros(dims))
+        val counts = Array.fill(k)(0L)
+
+        val numRows = pointMatrix.numRows
+
+        // Construct centers matrix
+        val centersArray = new Array[Double](k * dims)
+        val centersNormArray = new Array[Double](k)
+        var i = 0
+        thisCenters.foreach { center =>
+          System.arraycopy(center.vector.toArray, 0, centersArray, i * dims, dims)
+          centersNormArray(i) = math.pow(center.norm, 2.0)
+          i += 1
+        }
+        val centerMatrix = new DenseMatrix(dims, k, centersArray)
+
+        val a2b2 = new Array[Double](k * numRows)
```

Comment:
Should `a2b2` (or `distanceMatrix`) be persistent along with `pointMatrix` and `pointsNormMatrix`? It seems that it is created every iteration. There are also some other structures, like `sums`, that might be persisted too.